### ISSUE TITLE
New engine(?): Add support for Lingva (alternative front-end for Google Translate)

### ISSUE THREAD

**Issue by Vinfall** [soimort/translate-shell#462]
- [x] I've read the [contributing guidelines](https://github.com/soimort/translate-shell/blob/3dcc1229d98f60df695b9cc81fa5cdbd4d53cc63/CONTRIBUTING.md).
- [x] I'm using the latest stable version (0.9.7)

Is it possible to add support for [Lingva Translate](https://github.com/thedaviddelta/lingva-translate), or any alternative front-end for popular translate services in general?

Basically it will return just the same result except that the query is done on the remote server so users can retrieves the translation without directly accessing the respective translate service.

In Lingva's case, it offers a RESTful API and a GraphQL one which allow specify source/target language and have support for audio. You can check the details in [Public APIs](https://github.com/thedaviddelta/lingva-translate#public-apis) section of the project README.

For me, this issue serves as a general inquiry about your attitude towards these services as historically speaking, such projects commonly get abandoned after a while and their APIs are subject to change from time to time. Please feel free to add a *wontfix* lable if that's your response regarding this and thank you for your dedication to the project!

Edit: fix a typo.

**Comment by soimort** [comment]
Having RESTful and GraphQL APIs would be cool. But I do have several concerns:

1. If I understand correctly, Lingva serves like a middleware and does not provide anything extra (other than the more accessible API formats). If `trans` can already parse the API from Google's server, there is no need for employing an agent.
2. There is no guarantee that those Lingva instances will be maintained for the long term (and there is probably also a risk of violation of Google's ToS). Meanwhile I'll bet https://translate.google.com/ will always be there as long as Google doesn't shut it down.

So I guess since translate-shell is an alternative (command-line) front-end itself, it might not be super interesting to make it a front-end of another alternative front-end...

**Comment by Vinfall** [comment]
Yeah, I share the same concerns with you. The only benefit would be just not directly accessing Google's service (by requesting the alternative front-end server) as few people have no access to it due to poisoned DNS, corporate firewall etc, while at the same time, maintaining support for these middlewares may cost unnecessary effort.

I second the alternative *front-end* narrative and would like to point out that the main difference between translate-shell and Lingva is whether redirecting the request to Google's server. But still, the benefit is limited and the availability of these services is subject to change.

Thank you for sharing your thoughts on this and I'll leave it for you to decide whether to close the issue.

**Comment by Andrew15-5** [comment]
Hey there. I'm really enjoying the project and can't live without it, as shell is the fastest way to get anything done (change my mind). I'm trying to switch from Google to FOSS as much as possible, and next on my list is translation. I'm trying to use DeepL on Android, but it's kinda buggy with language selection/changing. But on my main machine, I'm usually only using `trans`. Today I found about [Dialect](https://github.com/dialect-app/dialect), but it's a GUI app, which maybe fine in the future when I test it more (3 s before I can write prompt). It has an option of using Lingva with any URL to the server. Now I think using Lingva is the best solution as I won't lose on the accuracy, but also won't use Google (directly, i.e., no tracking).

Now I see that the project doesn't yet support Lingva, which is a shame as it's really easy to use (at least the HTTP API, I don't speak GraphQL yet):

```sh
curl -s https://lingva.ml/api/v1/en/ja/test | jq -r .translation # テスト
```

I had a thought of maybe making a simple wrapper for this to use in terminal, but I think that it would be much better if translate-shell add this engine directly.

Is this feature in the backlog? Can I make a PR? I also noticed that `trans` can actually already speak (never used it, but I guess I should), so this means Ligva's support for audio should also work.

**Comment by Vinfall** [comment]
Oh, I almost forgot this issue.

Actually Lingva has been inactive for 2 years, which is also why it is not planned. And it only seems more unnecessary for now and I'm closing this.

FWIW, now I use [mozhi](https://codeberg.org/aryak/mozhi) instead of translate-shell, which supports both web ui (via a simple `mozhi serve` command) _and_ command line.

I also have shell rc to make it behave like translate-shell:
```sh
# shell alias, auto to zh-CN, like `trans :zh $1`
alias tz="mozhi translate --engine google --dest zh-CN --source auto --query $1"

# or function, auto-detect to en
t() {
    mozhi translate --engine google --dest en --source auto --raw --query "$1" | jq -r '.["translated-text"]'
}

# use
$ tz 'hello'
Translated Text: 你好
Detected Language: en
Source Language: auto
Target Language: zh-CN
$ t テスト
test
```

**Comment by Andrew15-5** [comment]
There are still instances that are live and anyone can host their own instance, so the fact that the project is not being updated doesn't mean it is bad or won't work.

The problem with mozhi and other tools, is that they are not as feature rich as this project and therefore won't be good enough for everyone, including myself. I usually use the verbose translation that shows many translations, which often can help find some word or choose a better translation (there is also a lot of complementary info). And the audio feature is also quite nice.

In the end, I can always make my own CLI, though currently I have different priorities. So, maybe one day.

**Comment by Vinfall** [comment]
I do agree with you and translate-shell written in AWK is super cool.

This is closed simply because I don't find it relevant for myself. You can probably open a new issue or even with a minimal showcase instead.
